File: src/go/radio/api_module.go
package radio

import (
	"context"
	"time"
)

// ApiModule API 子模块的基类。
type ApiModule struct {
	rm      *RadioManager
	nowMode string
}

func NewApiModule(rm *RadioManager) *ApiModule {
	return &ApiModule{
		rm:      rm,
		nowMode: "sync",
	}
}

func (am *ApiModule) Mode(mode string) *ApiModule {
	am.nowMode = mode
	return am
}

func (am *ApiModule) Send(cmd string, data map[string]interface{}, waitCmd string, timeout time.Duration, postProcessor func(map[string]interface{}) interface{}) interface{} {
	switch am.nowMode {
	case "token":
		return am.rm.SendWithToken(cmd, data, waitCmd, postProcessor)
	case "async":
		// Go 中异步通常通过返回 channel 或使用 context。
		// 这里为了简化，我们让用户自己处理 context。
		ctx, cancel := context.WithTimeout(context.Background(), timeout)
		defer cancel()
		res, _ := am.rm.SendAndWaitAsync(ctx, cmd, data, waitCmd, postProcessor)
		return res
	case "sync":
		fallthrough
	default:
		return am.rm.SendAndWaitSync(cmd, data, waitCmd, timeout, postProcessor)
	}
}

File: src/rs/radio/api_module.rs
use std::sync::{Arc, Weak};
use std::time::Duration;
use serde_json::Value;
use crate::radio::radio_manager::RadioManager;
use crate::radio::wait_token::WaitToken;

pub enum SendResult {
    Sync(Option<Value>),
    Token(Arc<WaitToken>),
}

pub struct ApiModule {
    pub rm: Weak<RadioManager>,
    pub mode: String, // "sync", "token"
}

impl ApiModule {
    pub fn new(rm: Weak<RadioManager>) -> Self {
        Self {
            rm,
            mode: "sync".to_string(),
        }
    }

    pub fn set_mode(&mut self, mode: &str) {
        self.mode = mode.to_string();
    }

    pub fn send(&self, cmd: &str, data: Option<Value>, wait_cmd: Option<&str>) -> SendResult {
        if let Some(rm) = self.rm.upgrade() {
            match self.mode.as_str() {
                "token" => {
                    let token = rm.send_with_token(cmd, data, wait_cmd);
                    SendResult::Token(token)
                },
                _ => {
                    let res = rm.send_and_wait_sync(cmd, data, wait_cmd, Duration::from_secs(3));
                    SendResult::Sync(res)
                }
            }
        } else {
            SendResult::Sync(None)
        }
    }
}

impl Default for ApiModule {
    fn default() -> Self {
        Self {
            rm: Weak::new(),
            mode: "sync".to_string(),
        }
    }
}
